ethcore/src/miner/mod.rs

```diff
@@ -17,191 +17,170 @@
+
+use block::{Block, SealedBlock};
 use client::{
-	MiningBlockChainClient, CallContract, RegistryInfo, ScheduleInfo,
-	BlockChain, AccountData, BlockProducer, SealedBlockImporter
+	CallContract, RegistryInfo, ScheduleInfo,
+	BlockChain, BlockProducer, SealedBlockImporter, ChainInfo,
+	AccountData, Nonce,
 };
-use error::{Error};
-use ethereum_types::{H256, U256, Address};
+use error::Error;
 use header::{BlockNumber, Header};
 use receipt::{RichReceipt, Receipt};
-use transaction::{UnverifiedTransaction, PendingTransaction, ImportResult as TransactionImportResult};
+use transaction::{self, UnverifiedTransaction, SignedTransaction, PendingTransaction};
 use state::StateInfo;
 
+/// Provides methods to verify incoming external transactions
+pub trait TransactionVerifierClient: Send + Sync
+	// Required for ServiceTransactionChecker
+	+ CallContract + RegistryInfo
+	// Required for verifiying transactions
+	+ BlockChain + ScheduleInfo + AccountData
+{}
+
+/// Extended client interface used for mining
+pub trait BlockChainClient: TransactionVerifierClient + BlockProducer + SealedBlockImporter {}
+
+// TODO [ToDr] Split into smaller traits?
+// TODO [ToDr] get rid of from_pending_block in miner/miner.rs
+
 /// Miner client API
 pub trait MinerService : Send + Sync {
 	/// Type representing chain state
 	type State: StateInfo + 'static;
 
-	/// Returns miner's status.
-	fn status(&self) -> MinerStatus;
-
-	/// Get the author that we will seal blocks as.
-	fn author(&self) -> Address;
+	// Sealing
 
-	/// Set the author that we will seal blocks as.
-	fn set_author(&self, author: Address);
-
-	/// Set info necessary to sign consensus messages.
-	fn set_engine_signer(&self, address: Address, password: String) -> Result<(), ::account_provider::SignError>;
-
-	/// Get the extra_data that we will seal blocks with.
-	fn extra_data(&self) -> Bytes;
-
-	/// Set the extra_data that we will seal blocks with.
-	fn set_extra_data(&self, extra_data: Bytes);
-
-	/// Get current minimal gas price for transactions accepted to queue.
-	fn minimal_gas_price(&self) -> U256;
+	/// Submit `seal` as a valid solution for the header of `pow_hash`.
+	/// Will check the seal, but not actually insert the block into the chain.
+	fn submit_seal(&self, pow_hash: H256, seal: Vec<Bytes>) -> Result<SealedBlock, Error>;
 
-	/// Set minimal gas price of transaction to be accepted for mining.
-	fn set_minimal_gas_price(&self, min_gas_price: U256);
+	/// Is it currently sealing?
+	fn is_currently_sealing(&self) -> bool;
 
-	/// Get the lower bound of the gas limit we wish to target when sealing a new block.
-	fn gas_floor_target(&self) -> U256;
+	/// Get the sealing work package preparing it if doesn't exist yet.
+	///
+	/// Returns `None` if engine seals internally.
+	fn work_package<C>(&self, chain: &C) -> Option<(H256, BlockNumber, u64, U256)>
+		where C: BlockChain + CallContract + BlockProducer + SealedBlockImporter + Nonce + Sync;
```

Comment:
I thought, we wanted to get rid of too granular traits like `Nonce`, so it would probably be better to bound on aggregate traits like `AccountData` here.